src/server/util/insertData.js

```diff
@@ -20,26 +20,51 @@ const fs = require('fs').promises;
 async function insertUnits(unitsToInsert, update = false, conn) {
 	await Promise.all(unitsToInsert.map(
 		async unitData => {
-			const dbUnit = await Unit.getByName(unitData[0], conn);
-			if (dbUnit === null) {
-				// The unit does not exist so add it.
-				await new Unit(undefined, unitData[0], unitData[1], unitData[2], unitData[3],
-					unitData[4], null, unitData[5], unitData[6], unitData[7], unitData[8]).insert(conn);
-			} else if (update) {
-				// Asked to update so will. Does not bother to check if no changes.
-				dbUnit.name = unitData[0];
-				dbUnit.identifier = unitData[1];
-				dbUnit.unitRepresent = unitData[2];
-				dbUnit.secInRate = unitData[3];
-				dbUnit.typeOfUnit = unitData[4];
-				dbUnit.suffix = unitData[5];
-				dbUnit.displayable = unitData[6];
-				dbUnit.preferredDisplay = unitData[7];
-				dbUnit.note = unitData[8];
-				// Should update even though exists.
-				await dbUnit.update(conn);
+			let ok = true;
+			for (let i = 0; i < unitsToInsert.length; ++i) {
+				// Meter key/value pairs for the current meter.
+				const uData = unitsToInsert[i];
+		
+				// Check that needed keys are there.
+				const requiredKeys = ['name', 'unitRepresent', 'typeOfUnit', 'displayable', 'preferredDisplay'];
+				ok = true;
+				requiredKeys.forEach(key => {
+					if ((key == 'name') && (typeof uData[key] !== 'string')){
```

Comment:
A few thoughts on these new checks:

1. The formatting is off. I suggest you use the standard Visual Studio Code formatting if you are using that IED.
2. These checks differ from the way meter does it that was given as an example. I can understand that checking the type of the value could be valuable. However, we have not been doing that and only doing in some cases seems inconsistent. It might be easiest to remove them as there are lots of ways that the data could be incorrect. If you want to keep this then some structure of requiredKeys that has the name and type that is then checked the same way for all keys seems better. You should also update meters to be consistent.
3. The first and second tests seem to do the same thing but are separated out. I'm unclear why. This will go away if you replace with the way meter did it or if you do consistent checks for all keys.

Reply:
Understood. The checks differ from the way meters does it because some nonstring keys return values such as 0 or false which will cause the program to think the key is missing. I could only think of the basic workaround which is to check the type of value, if there is any suggestions on how to do these checks better or change the checks meters does to fit units and conversions please let me know.

Reply:
While trying to figure out how to check the keys in a better way I ran into issues I found in the original code. I believe the issue was not seen because the same conversions were being inserted many times but the code skipped the insert after the first time. Below is my revised code to insert conversions that I believe is correct. The same basic logic should be used for units. The improved log message should be used in meters but it is important not to use a map function in that case so only change the message. You are welcome to make these changes or I can do them if you prefer. 

```js
async function insertConversions(conversionsToInsert, conn) {
	await Promise.all(await Promise.all(conversionsToInsert.map(
		async (conversionData, index) => {
			// Check that needed keys are there.
			const requiredKeys = ['sourceName', 'destinationName', 'bidirectional', 'slope', 'intercept'];
			ok = true;
			requiredKeys.forEach(key => {
				if (!conversionData.hasOwnProperty(key)) {
					console.log(`********key "${key}" is required but missing so conversion number ${index} not processed with values:`, conversionData);
					// Don't insert
					ok = false;
				}
			})
			if (ok) {
				const sourceName = (await Unit.getByName(conversionData.sourceName, conn)).id;
				const destinationName = (await Unit.getByName(conversionData.destinationName, conn)).id;
				if (await Conversion.getBySourceDestination(sourceName, destinationName, conn) === null) {
					await new Conversion(sourceName, destinationName, conversionData.bidirectional, conversionData.slope, conversionData.intercept, conversionData.note).insert(conn);
				}
			}
		}
	)))
}
```

Reply:
Sorry for the delay. I have updated insertConversions to match theses changes.